File: C++/Grind75/Week-1/733.flood-fill.cpp
#include <bits/stdc++.h>
using namespace std;
/*
 * @lc app=leetcode id=733 lang=cpp
 *
 * [733] Flood Fill
 */

// @lc code=start
class Solution
{
public:
    vector<vector<int>> floodFill(vector<vector<int>> &image, int sr, int sc, int color)
    {
        int ocolor = image[sr][sc];

        if (ocolor == color)
        {
            return image;
        }

        image[sr][sc] = color;

        if (!(sr + 1 >= image.size()) && image[sr + 1][sc] == ocolor)
        {
            image = floodFill(image, sr + 1, sc, color);
        }
        if (!(sc + 1 >= image[0].size()) && image[sr][sc + 1] == ocolor)
        {
            image = floodFill(image, sr, sc + 1, color);
        }
        if (sr - 1 >= 0 && image[sr - 1][sc] == ocolor)
        {
            image = floodFill(image, sr - 1, sc, color);
        }
        if (sc - 1 >= 0 && image[sr][sc - 1] == ocolor)
        {
            image = floodFill(image, sr, sc - 1, color);
        }
        return image;
    }
};
// @lc code=end

File: Java/Grind75/Week1/733.flood-fill.java
package Grind75.Week1;

/*
 * @lc app=leetcode id=733 lang=java
 *
 * [733] Flood Fill
 */

// @lc code=start
class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        int oColor = image[sr][sc];
        if (color == oColor)
            return image;
        image[sr][sc] = color;
        int[] dir = { 1, 0, -1, 0, 1 };
        for (int i = 1; i < dir.length; i++) {
            int nr = sr + dir[i - 1];
            int nc = sc + dir[i];
            if (nr >= 0 && nr < image.length && nc >= 0 && nc < image[0].length && image[nr][nc] == oColor)
                floodFill(image, nr, nc, color);
        }
        return image;
    }
}
// @lc code=end
